crates/subspace-service/src/piece_cache.rs

```diff
@@ -35,15 +46,56 @@ where
     const KEY_PREFIX: &[u8] = b"piece_cache";
 
     /// Create new instance with specified size (in bytes)
-    pub fn new(aux_store: Arc<AS>, cache_size: u64) -> Self {
+    pub fn new(aux_store: Arc<AS>, cache_size: u64, local_peer_id: PeerId) -> Self {
         let max_pieces_in_cache: PieceIndex = cache_size / PIECE_SIZE as PieceIndex;
+        let local_provided_keys = Self::get_local_provided_keys(aux_store.clone())
+            .expect("DB loading should succeed.")
+            .unwrap_or_default();
+
+        if local_provided_keys.is_empty() {
+            info!("New storage provider initialized.");
+        } else {
+            info!(
+                "Storage provider cache loaded - {} items.",
+                local_provided_keys.len()
+            );
+        }
 
         Self {
             aux_store,
             max_pieces_in_cache,
+            local_peer_id,
+            local_provided_keys,
         }
     }
 
+    fn get_local_provided_keys(
+        aux_store: Arc<AS>,
+    ) -> Result<Option<BTreeSet<PieceIndex>>, Box<dyn Error>> {
+        Ok(aux_store.get_aux(LOCAL_PROVIDED_KEYS)?.map(|data| {
+            let collection: ParityDbKeyCollection =
+                data.try_into().expect("DB loading should succeed.");
+
+            collection.set
+        }))
+    }
+
+    fn put_local_provided_keys(&self) -> Result<(), Box<dyn Error>> {
```

Comment:
More like `write_` or `persist_`. `put_` usually takes an argument of what is being put in.